agent/token_store.go

```diff
@@ -0,0 +1,84 @@
+package agent
+
+import (
+	"sync"
+)
+
+// TokenStore is used to hold the special ACL tokens used by Consul agents. It
+// is designed to update the tokens on the fly, so the token store itself should
+// be plumbed around and used to get tokens at runtime, don't save the resulting
+// tokens.
+type TokenStore struct {
+	// l synchronizes access to the token store.
+	l sync.RWMutex
+
+	// userToken is passed along for requests when the user didn't supply a
+	// token, and may be left blank to use the anonymous token. This will
+	// also be used for agent operations if the agent token isn't set.
+	userToken string
+
+	// agentToken is used for internal agent operations like self-registering
+	// with the catalog and anti-entropy, but should never be used for
+	// user-initiated operations.
+	agentToken string
+
+	// agentMasterToken is a special token that's only used locally for
+	// access to the /v1/agent utility operations if the servers aren't
+	// available.
+	agentMasterToken string
+}
+
+// UpdateUserToken replaces the current user token in the store.
+func (t *TokenStore) UpdateUserToken(userToken string) {
+	t.l.Lock()
+	t.userToken = userToken
+	t.l.Unlock()
+}
+
+// UpdateAgentToken replaces the current agent token in the store.
+func (t *TokenStore) UpdateAgentToken(agentToken string) {
+	t.l.Lock()
+	t.agentToken = agentToken
+	t.l.Unlock()
+}
+
+// GetTokenForUser returns the best token to use for user operations.
+func (t *TokenStore) GetTokenForUser() string {
+	t.l.RLock()
+	defer t.l.RUnlock()
+
+	if t.userToken != "" {
+		return t.userToken
+	}
+	return ""
+}
+
+// GetTokenForAgent returns the best token to use for internal agent operations.
+func (t *TokenStore) GetTokenForAgent() string {
+	t.l.RLock()
+	defer t.l.RUnlock()
+
+	if t.agentToken != "" {
+		return t.agentToken
+	}
+	if t.userToken != "" {
+		return t.userToken
+	}
+	return ""
+}
+
+// UpdateAgentMasterToken replaces the current agent master token in the store.
+func (t *TokenStore) UpdateAgentMasterToken(agentMasterToken string) {
```

Comment:
Haha I had intentionally put it near the other master token method. I'll move it up though since that's probably more natural :-)